contracts/core/GlobalReentrancyLock.sol

```diff
@@ -0,0 +1,237 @@
+pragma solidity 0.8.13;
+
+import {PermissionsV2} from "./PermissionsV2.sol";
+import {IGlobalReentrancyLock} from "./IGlobalReentrancyLock.sol";
+
+/// @notice inpsired by the openzeppelin reentrancy guard smart contracts
+/// data container size has been changed.
+
+/// @dev allows contracts and addresses with the LOCKER_ROLE to call
+/// in and lock and unlock this smart contract.
+/// once locked, only the original caller that locked can unlock the contract
+/// without the governor emergency unlock functionality.
+/// Governor can unpause if locked but not unlocked.
+
+/// @notice explanation on data types used in contract
+
+/// @dev block number can be safely downcasted without a check on exceeding
+/// uint88 max because the sun will explode before this statement is true:
+/// block.number > 2^88 - 1
+/// address can be stored in a uint160 because an address is only 20 bytes
+
+/// @dev in the EVM. 160bits / 8 bits per byte = 20 bytes
+/// https://docs.soliditylang.org/en/develop/types.html#address
+
+abstract contract GlobalReentrancyLock is IGlobalReentrancyLock, PermissionsV2 {
+    /// -------------------------------------------------
+    /// -------------------------------------------------
+    /// ------------------- Constants -------------------
+    /// -------------------------------------------------
+    /// -------------------------------------------------
+
+    uint8 private constant _NOT_ENTERED = 0;
+    uint8 private constant _ENTERED_LEVEL_ONE = 1;
+    uint8 private constant _ENTERED_LEVEL_TWO = 2;
+
+    /// ------------- System States ---------------
+
+    /// system unlocked
+    /// request level 2 locked
+    /// level 2 locked, msg.sender stored
+    /// level 2 unlocked, msg.sender checked to ensure same as locking
+    /// unlock down to level 0
+    ///
+    /// system unlocked
+    /// request level 1 locked
+    /// level 1 locked, msg.sender stored
+    /// level 1 unlocked, msg.sender checked to ensure same as locking
+    ///
+    /// lock level 1, msg.sender is stored
+    /// request level 2 locked
+    /// level 2 locked, msg.sender not stored
+    /// level 2 unlocked, msg.sender not checked
+    /// level 1 unlocked, msg.sender checked
+    ///
+    /// level 1 locked
+    /// request level 2 locked
+    /// level 2 locked
+    /// request level 0 unlocked, invalid state, must unlock to level 1, call reverts
```

Comment:
Make another pass on the comments in this file since we changes the logic to "lock one level above, unlock one level below only"

Reply:
done